### ISSUE TITLE
Deletion of PVs when cluster is deleted for AWS EBS volumes

### ISSUE THREAD

**Issue by kunal-shah** [kubernetes/kubernetes#55632]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug

I am using dynamic provisioning of storage with AWS EBS. When I delete Kubernetes cluster, all physical volumes on EBS that were created are still present and not getting deleted. I was expecting them to be deleted with cluster that is managing those volumes. If entire cluster is gone, there is no good way to re-utilize those dynamically created EBS volumes anyway. 
I found the following comment in the code for AWS management:
// TODO: Should we tag this with the cluster id (so it gets deleted when the cluster does?)
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L138
So looks like this is not implemented. Is my understanding correct and is there a plan to add this? 

**What happened**:
EBS volumes do not get deleted on deletion of kubernetes cluster. 

**What you expected to happen**:
EBS volumes are tagged by KubernetesCluster name but they don't get deleted explicitly. 

**How to reproduce it (as minimally and precisely as possible)**:
Create dynamically provisioned PVs using AWS EBS. Delete kubernetes cluster. EBS volumes are still present.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by gnufied** [comment]
How did you delete your Cluster? Deleting a namespace should cause deletion of all dynamically provisioned PVs and their backing EBS volumes.

**Comment by kunal-shah** [comment]
Thanks for your prompt response. I delete AWS stack that created kubernetes
cluster and hence everything gets deleted. What triggers namespace
deletion?

Undeploying my pods and resources doesn't impact PVs which is expected I
guess. Any particular CLI I can use to trigger this?

Thanks,
Kunal

**Comment by kunal-shah** [comment]
Hi Guys, I will appreciate any help here. I am kind of stuck at this point
and looks like cannot use dynamic provisioning for my use case (creating
AWS stack with kubernetes cluster, run load test and destroy stack with all
resources including EBS volumes).

**Comment by gnufied** [comment]
You should just create a new namespace (and not use `default` namespace) and run your load test in it. Before deleting the cluster, just delete the namespace (via `kubectl delete namespace foo`) and that will delete all dynamically provisioned EBS volumes.

I am not sure what mechanism you are using for deploying and destoying AWS cluster. Is it `kops`? In that case - perhaps kops should delete namespaces before deleting the cluster.

**Comment by kunal-shah** [comment]
Thanks Hemant for your suggestions.. We are currently using heptio
cloudformation templates to create kubernetes cluster.  Destroying is just
done directly using cloudformation stack deletion. i will look into
creating cluster in a separate namespace.

Meanwhile, can you please tell me if there is a way to add custom AWS tags
to storageclass? We need all AWS resources to have tags corresponding to
which project they belong to, but based on storageclass definition, i can't
find any way to provide these custom tags (labels doesn't seem to be used
for this purpose).

Thanks,
Kunal

**Comment by gnufied** [comment]
I did not mean create cluster in separate namespace.  I am talking about kube namespaces - https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/

**Comment by kunal-shah** [comment]
Yep understood.. "Cluster" was a wrong word.. It should be all my pod
deployments.. Thanks for clarifying. Can you also help with adding custom
tags? Is it possible to do today with AWS and PVs?
